### DISCUSSION TITLE
Generel questions MQTT & GPIOs

### DISCUSSION THREAD

**Discussion by electrontube** [rednblkx/HomeKey-ESP32#86] [General]
First of all, a big thank you for the project and the initiative! 

Somehow I don't quite understand it yet...

Using Apple devices for authentication via NFC (instead of RFiD tags) is basically a good approach.

What I think I have understood:
The ESP32 takes care of the communication with the Apple devices and sends various data via MQTT to the “logic layer”

All decisions should be made in the logic layer and actions should be triggered from there via MQTT (or via other channels) on the actuators.

From my point of view, this would require the ability to control some GPIOs on the reader ESP32 via MQTT, e.g. acoustic feedback, triggering an electric door opener, etc....

Unfortunately, the existing documentation does not answer my questions. There are various MQTT topics in the system, but it is not clear to me which of them are published and which can be subscribed to and how they can then be linked  to GPIO outputs.

Perhaps someone can shed some light on this.

In addition to the possibility of controlling GPIOs via MQTT, I am also missing some kind of password protection for the web interface.

Thanks again and keep up the good work!

**Comment by rednblkx** [comment]
Hi!👋 

I’ll starting by saying this project was never meant to be hardware-focused, for this I believe something like ESPHome would be more suitable and I’m actively working to integrate HomeKit and HomeKey into ESPHome.

Apart from some basic HIGH/LOW on one GPIO pin controlled from the HomeKit entity or on successful Homekey there is no GPIO output control mechanism.

I’m sorry about the mess in the wiki, planning to overhaul parts of it, especially the MQTT part.

The only topics that you can subscribe to are `MQTT_STATE_TOPIC`, `MQTT_AUTH_TOPIC` and `MQTT_CUSTOM_STATE_TOPIC`. The other topics are control topics used to set the lock state.

> I am also missing some kind of password protection for the web interface.

I’ve received some other requests regarding this, I understand the need for this and it will be added in a future version.

**Reply by maslyankov** [reply]
Thank you so much for the effort! Having this functionality inside of ESPHome would be gamechanger!
